ranking/src/ts_system/mod.rs

```diff
@@ -0,0 +1,314 @@
+                    RefCell::borrow_mut(&players.last().unwrap().1.upgrade().unwrap()).0 =
+                        self.rating.get(&players.last().unwrap().0).unwrap().clone();
+
+                    let mut tmp: Vec<&mut dyn ValueNode> = Vec::with_capacity(3);
+                    tmp.push(&mut p[i][j][k]);
+                    tmp.push(&mut s[i][j][k]);
+                    tmp.push(&mut perf[i][j][k]);
+                    sp.push(SumNode::new(&mut tmp));
+                    RefCell::borrow_mut(perf[i][j][k].get_edges_mut().last_mut().unwrap()).1 = Gaussian { mu: 0., sigma: self.beta };
+                }
+
+                let mut tt: Vec<&mut dyn ValueNode> = vec![&mut t[i][j]];
+                for pp in &mut p[i][j] {
+                    tt.push(pp);
+                }
+                pt.push(SumNode::new(&mut tt));
+                let mut tmp: Vec<&mut dyn ValueNode> = Vec::with_capacity(3);
+                tmp.push(&mut l[i]);
+                tmp.push(&mut t[i][j]);
+                tmp.push(&mut u[i][j]);
+                tul.push(SumNode::new(&mut tmp));
+                conv.push(t[i][j].get_edges().last().unwrap().clone());
+            }
+
+            if i != 0 {
+                let mut tmp: Vec<&mut dyn ValueNode> = Vec::with_capacity(3);
+                let (a, b) = l.split_at_mut(i);
+                tmp.push(a.last_mut().unwrap());
+                tmp.push(b.first_mut().unwrap());
+                tmp.push(&mut d[i - 1]);
+                ld.push(SumNode::new(&mut tmp));
+            }
+        }
+
+        infer3(&mut s);
+        infer1(&mut sp);
+        infer3(&mut p);
+        infer1(&mut pt);
+        infer2(&mut t);
+        infer1(&mut tul);
+        infer2(&mut u);
+        infer1(&mut tul);
+
+        let mut rounds = 0;
+
+        while check_convergence(&conv, &old_conv) >= self.convergence_eps {
+            old_conv.clear();
+            for item in &conv {
+                old_conv.push(RefCell::borrow(item).clone());
+            }
+            rounds += 1;
+
+            infer_ld(&mut ld, &mut l);
+            infer1(&mut d);
+            infer_ld(&mut ld, &mut l);
+            infer1(&mut tul);
+            infer2(&mut u);
+            infer1(&mut tul);
+        }
+
+        eprintln!("Rounds until convergence: {}", rounds);
+
+        infer2(&mut t);
+        infer1(&mut pt);
+        infer3(&mut p);
+        infer1(&mut sp);
+        infer3(&mut s);
+
+        for (name, mess) in &players {
+            let prior;
+            let performance;
+
+            prior = RefCell::borrow(&Weak::upgrade(mess).unwrap()).0.clone();
+            performance = RefCell::borrow(&Weak::upgrade(mess).unwrap()).1.clone();
+
+            *self.rating.get_mut(name).unwrap() = prior * performance;
+        }
+    }
+}
+
+impl RatingSystem for TrueSkillSPBSystem {
+    fn round_update(&mut self, mut standings: Vec<(&mut Player, usize, usize)>) {

```

Comment:
I'd like to understand better why this became `&mut self`, and if there's any way around it.